ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/GeoIspLookup.java

```diff
@@ -0,0 +1,167 @@
+  }
+
+  @VisibleForTesting
+  static synchronized void clearSingletonsForTests() {
+    singletonIspReader = null;
+  }
+
+  private static synchronized DatabaseReader getOrCreateSingletonIspReader(
+      ValueProvider<String> ispDatabase) throws IOException {
+    if (singletonIspReader == null) {
+      File mmdb;
+
+      try {
+        InputStream inputStream = BeamFileInputStream.open(ispDatabase.get());
+        Path mmdbPath = Paths.get(System.getProperty("java.io.tmpdir"), "GeoIspLookup.mmdb");
+        Files.copy(inputStream, mmdbPath, StandardCopyOption.REPLACE_EXISTING);
+        mmdb = mmdbPath.toFile();
+      } catch (IOException e) {
+        throw new IOException("Exception thrown while fetching configured geoIspDatabase", e);
+      }
+      singletonIspReader = new DatabaseReader.Builder(mmdb).withCache(new CHMCache()).build();
+    }
+    return singletonIspReader;
+  }
+
+  @VisibleForTesting
+  public class Fn extends SimpleFunction<PubsubMessage, PubsubMessage> {
+
+    private transient DatabaseReader ispReader;
+
+    private final Counter foundIp = Metrics.counter(GeoCityLookup.Fn.class, "found_ip");
+    private final Counter countIspAlreadyApplied = Metrics.counter(Fn.class, "isp_already_applied");
+    private final Counter foundIsp = Metrics.counter(Fn.class, "found_isp");
+    private final Counter countIpForwarded = Metrics.counter(GeoCityLookup.Fn.class,
+        "ip_from_x_forwarded_for");
+    private final Counter countIpRemoteAddr = Metrics.counter(GeoCityLookup.Fn.class,
+        "ip_from_remote_addr");
+
+    @Override
+    public PubsubMessage apply(PubsubMessage message) {
+      message = PubsubConstraints.ensureNonNull(message);
+
+      try {
+        if (ispReader == null) {
+          loadResourcesOnFirstMessage();
+        }
+
+        if (message.getAttributeMap().containsKey(Attribute.ISP_NAME)) {
+          // Return early since ISP lookup has already been performed.
+          countIspAlreadyApplied.inc();
+          return message;
+        }
+
+        // copy attributes
+        Map<String, String> attributes = new HashMap<String, String>(message.getAttributeMap());
+
+        // Determine client ip
+        String ip;
+        String xff = attributes.get(Attribute.X_FORWARDED_FOR);
+
+        if (xff != null) {
+          // Google's load balancer will append the immediate sending client IP and a global
+          // forwarding rule IP to any existing content in X-Forwarded-For as documented in:
+          // https://cloud.google.com/load-balancing/docs/https/#components
+          //
+          // In practice, many of the "first" addresses are bogus or internal,
+          // so we target the immediate sending client IP.
+          String[] ips = xff.split("\\s*,\\s*");
+          ip = ips[Math.max(ips.length - 2, 0)];
+          countIpForwarded.inc();
+        } else {
+          ip = attributes.getOrDefault(Attribute.REMOTE_ADDR, "");
+          countIpRemoteAddr.inc();
+        }
```

Comment:
I'd really like to avoid repeating this code for determining IP. We should break this up into:

```
            .apply(options.getInputType().read(options)) //
            .apply(ParseProxy.of()) // This would also now have responsibility for outputting an "ip" attribute
            .apply(GeoIspLookup.of(options.getGeoIspDatabase())) //
            .apply(GeoCityLookup.of(options.getGeoCityDatabase(), options.getGeoCityFilter())) // This would no longer strip IP address attributes
            .apply(DropIP.of()) // new transform that drops all IP address attributes
```

That probably should be a separate PR, though. This is fine for now.